apps/settings/js/panels/search/search.js

```diff
@@ -1,79 +1,25 @@
 'use strict';
 
 define(function(require) {
-  var SettingsCache = require('modules/settings_cache');
-  var LazyLoader = require('shared/lazy_loader');
+
+  var SearchProvider = require('shared/search_provider');
 
   function Search() {
-    this._searchUrlTemplate = null;
     this._searchEngineSelect = null;
   }
 
   Search.prototype.init = function(searchEngineSelect) {
     this._searchEngineSelect = searchEngineSelect;
-    this.getCurrentSearchEngine();
-    this.initSearchEngineSelect();
-  };
-
-  Search.prototype.getCurrentSearchEngine = function() {
-    SettingsCache.getSettings((function(settingsCache) {
-      this._searchUrlTemplate = settingsCache['search.urlTemplate'];
-    }).bind(this));
-  };
-
-  Search.prototype.initSearchEngineSelect = function() {
-    // Get search provider list from settings
-    SettingsCache.getSettings((function(settingsCache) {
-      var searchEngineList = settingsCache['search.providers'];
-      // If the list is empty, populate it from default JSON file
-      if (!searchEngineList) {
-        this.populateSearchEngines()
-        .then(this.generateSearchEngineOptions.bind(this));
-        return;
-      }
-
-      // Otherwise just go ahead and generate the options
-      this.generateSearchEngineOptions(searchEngineList);
-    }).bind(this));
-  };
-
-  /**
-   * Populate search engine list in settings from default JSON file.
-   *
-   * @param {Function} callback function to call with retrieved data.
-   */
-  Search.prototype.populateSearchEngines = function() {
-    return LazyLoader.getJSON('/resources/search/providers.json')
-    .then(function(data) {
-      if (!data) {
-        return;
-      }
-
-      var result = navigator.mozSettings.createLock().set({
-        'search.providers': data
-      });
-      result.onerror = function() {
-        console.error('Unable to set search providers setting');
-      };
-
-      return data;
-    })
-    .catch(function () {
-      console.error('Unable to get default search provider file.');
-      return;
-    });
+    SearchProvider.providerUpdated(this.drawProviders.bind(this));
+    this.drawProviders();
```

Comment:
There would be chanced that at this moment, `SearchProvider.providers();` would retunrn `{}`.

For `SearchProvider`, it should provide a mechanism that callee can definitely get wanted provider information without error. So, this line should be something like this : 

```javascript
SearchProvider.ready().then(() => {
  this.drawProviders()
});
```